### PR TITLE
Centralize creating headers and sending data TO the server

### PR DESCRIPTION
**Pull request by designatednerd**
A follow-up to #848 inspired by some weird shit I was seeing prepping to adjust push notification handling for network switching. Centralizes creation of signed headers and doing `POST` and `PUT` requests (and one oddball `GET` request).

### CODE REVIEW

File: Toshi/APIClients/IDAPIClient.swift [hunk 1]
```diff
@@ -157,9 +157,11 @@ final class IDAPIClient {
                 return
             }
 
-            self.fetchTimestamp { timestamp, error in
+            self.fetchTimestamp { timestamp, _ in
```
Comment (marijnschilling) [inline Toshi/APIClients/IDAPIClient.swift]: No more error? 😞

Comment (marijnschilling) [reply Toshi/APIClients/IDAPIClient.swift]: and: self? I guess that was already there though.

File: Toshi/APIClients/IDAPIClient.swift [hunk 2]
```diff
@@ -195,7 +194,7 @@ final class IDAPIClient {
                         guard response.statusCode == 200 else { return }
 
                         guard let data = json?.data else {
-                           assertionFailure("No data from registration request response")
+                            assertionFailure("No data from registration request response")
```
Comment (marijnschilling) [inline Toshi/APIClients/IDAPIClient.swift]: Just for my information: why can we throw an assertionFailure in this case and not just an error?

File: Toshi/APIClients/ToshiError.swift [hunk 3]
```diff
@@ -78,4 +78,21 @@ extension ToshiError {
         let validErrorDescription = errorDescription ?? teapotError.description
         self.init(withType: errorType, description: validErrorDescription, responseStatus: teapotError.responseStatus, underlyingError: teapotError.underlyingError)
     }
+
+    init(errorResult: NetworkResult) {
```
Comment (marijnschilling) [inline Toshi/APIClients/ToshiError.swift]: 😍

File: Toshi/APIClients/ChatAPIClient.swift [hunk 4]
```diff
@@ -54,50 +53,37 @@ final class ChatAPIClient {
...
                         succeeded = true
-                    case .failure(_, _, let error):
-                        DLog("\(error)")
-                        succeeded = false
-                    }
-
-                    DispatchQueue.main.async {
-                        completion(succeeded)
+                    case .failure:
+                        // Nothing to do here except what's in the defer block
```
[the remark this answers is not in the document]
Comment (designatednerd) [reply Toshi/APIClients/ChatAPIClient.swift]: Hm...fair on the comment. Would it be better if I just deleted the comment?

File: Toshi/APIClients/ToshiError.swift (hunk 3 above)
Comment (Elland) [reply Toshi/APIClients/ToshiError.swift]: 👏🏽 now that's a great idea!

**Pull request merged**